### PR TITLE
Integrate the test build file into the main build file

### PR DESCRIPTION
**Pull request by aminosbh**
This modification adapts the build file of the test binary to the new build file of the library.
To build the test binary we just have to enable the option BUILD_TESTS like the following:
cmake .. -DBUILD_TESTS=ON
You can then update directly the pull request to DavidLeeds :wink:

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 cmake_minimum_required(VERSION 3.5)
 project(hashmap VERSION 1.0.0 LANGUAGES C)
 
+# Option to build the hashmap tests
+option(BUILD_TESTS "Build tests" OFF)
```
Comment (tarik02) [inline CMakeLists.txt:5]: Why do we need BUILD_TESTS option? At least, if we've it, it should be turned ON by default

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,10 @@
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 3.5)
+project(hashmap_test VERSION 1.0.0 LANGUAGES C)
```
Comment (tarik02) [inline test/CMakeLists.txt:2]: I think, tests shouldn't be a separate project, tests are just another executable, but absolutely not another project

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wunused -Werror")
+add_executable(hashmap_test
+    hashmap_test.c
+)
 
-# Build hashmap library with metrics enabled
-add_definitions(-DHASHMAP_METRICS)
+target_compile_options(hashmap_test PRIVATE $<$<CXX_COMPILER_ID:GNU>:-Wall -Wunused -Werror>)
 
-include_directories(../src)
```
Comment (tarik02) [inline test/CMakeLists.txt:8]: There's a problem. Let's assume, hashmap has some dependency (library). Assume that it's header (hashmap.h) includes it. We will get an error here.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
-
-add_executable(hashmap_test ../src/hashmap.c hashmap_test.c)
```
Comment (tarik02) [inline test/CMakeLists.txt:10]: You shouldn't add `hashmap.c` to the executable

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+target_link_libraries(hashmap_test HashMap::HashMap)
```
Comment (tarik02) [inline test/CMakeLists.txt:10]: It's enough to just add link libraries to add library. It automatically adds static/shared library and include directories.

**Review by tarik02 — CHANGES_REQUESTED** [review]
CMake is very complex thing, I can't understand couldn't they make it much simpler... maybe I am wrong.

Look here for test example: https://github.com/pabloariasal/modern-cmake-sample/blob/master/libjsonutils/test/CMakeLists.txt
And read this article: https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/

File: CMakeLists.txt (hunk 1 above)
Comment (aminosbh) [reply CMakeLists.txt:5]: Actually, in the example of Pablo Arias (I liked the article), he is using the GoogleTest (Specially dedicated for C++ code) to execute the tests. But in our case David Leeds just wrote a custom binary in C to execute some performance tests in his own manner which is grate. So I just simplified the build file to directly build the tests without GoogleTest.
The reason why I added an option for building tests is the additional flag (HASHMAP_METRICS) which is required to build the tests (See the original build file).
To summarize, when the option BUILD_TESTS is ON, we build the library with the flag HASHMAP_METRICS, we build the test binary and we link it with the library (built with the HASHMAP_METRICS flag). And when the BUILD_TESTS is OFF, we just build the library without the HASHMAP_METRICS flag. So the option BUILD_TESTS is necessary.

File: test/CMakeLists.txt (hunk 2 above)
Comment (aminosbh) [reply test/CMakeLists.txt:2]: I don't know, may be. But I considered the test binary as a sub-project of library.

File: test/CMakeLists.txt (hunk 3 above)
Comment (aminosbh) [reply test/CMakeLists.txt:8]: In the code of Pablo Arias, he wrote the build file of the test in a manner to be build as separate project that uses the library. I'm sure that you remarked that he didn't integrated the test as a subdirectory in the library build file but the library as a subdirectory of the test. In fact, I'm sure that his reason is all about the HASHMAP_METRICS flag that should be enabled when building the test.

File: test/CMakeLists.txt (hunk 4 above)
Comment (aminosbh) [reply test/CMakeLists.txt:10]: Yes, we shouldn't :wink: I just removed it.

File: test/CMakeLists.txt (hunk 5 above)
Comment (aminosbh) [reply test/CMakeLists.txt:10]: Yes, I confirm :wink:

File: CMakeLists.txt (hunk 1 above)
Comment (tarik02) [reply CMakeLists.txt:5]: Okay, I understand. If so, we should build two variants of hashmap (two separate targets):
 - `hashmap`
 - `hashmap_metrics`
The first one is just normal build. The second one is hashmap built with `-DHASHMAP_METRICS`.

Also, I didn't say anything about GoogleTest. I mention only three last lines of that file. So, there should be two options:
 - `BUILD_METRICS` "Build library hashmap_metrics"
 - `BUILD_TESTS` "Build tests"

I think, `BUILD_TESTS` should be enabled by default. Also, if `BUILD_TESTS` is enabled, then `BUILD_METRICS` should be forcefully enabled.

This approach should be flexible enough (users can use metrics library if they need metrics).

File: test/CMakeLists.txt (hunk 2 above)
Comment (tarik02) [reply test/CMakeLists.txt:2]: Tests shouldn't be separate project at least because tests do not have separate versioning.

File: CMakeLists.txt (hunk 1 above)
Comment (aminosbh) [reply CMakeLists.txt:5]: Okay, It a good idea to create two separate targets (hashmap and hashmap_metric) so that we can enable building test by default. But the hashmap_metrics should be just used by the test binary so no need to export it like the "hashmap" (No need for a custom option for it). Actually, the flag HASHMAP_METRICS is just used in conjunction with the test. So normal users (including developer) will not use it (except to make contribution to this project, so build from source).

File: test/CMakeLists.txt (hunk 2 above)
Comment (aminosbh) [reply test/CMakeLists.txt:2]: I approve.

File: CMakeLists.txt (hunk 1 above)
Comment (aminosbh) [reply CMakeLists.txt:5]: I tried to find a way to clone the hashmap target but I didn't find a clean way to do it. Two solutions are offered to me : 
- Copy the whole hashmap config (source, include, compile options) manually.
- Copy the whole hashmap config dynamically using the following code : 
```cmake
if(BUILD_TESTS)
    # Clone hashmap library properties
    get_target_property(HASHMAP_SOURCES hashmap SOURCES)
    get_target_property(HASHMAP_INCLUDE_DIRECTORIES hashmap INCLUDE_DIRECTORIES)
    get_target_property(HASHMAP_COMPILE_OPTIONS hashmap COMPILE_OPTIONS)
    add_library(hashmap_metrics ${HASHMAP_SOURCES})
    target_include_directories(hashmap_metrics PUBLIC ${HASHMAP_INCLUDE_DIRECTORIES})
    target_compile_options(hashmap_metrics PRIVATE ${HASHMAP_COMPILE_OPTIONS})

    # Add HASHMAP_METRICS flag
    target_compile_definitions(hashmap_metrics PUBLIC HASHMAP_METRICS)
endif()
```

Comment (tarik02) [reply CMakeLists.txt:5]: So... Then we need to put sources (and compile flags) to variables (`HASHMAP_SOURCES`, `HASHMAP_COMPILE_FLAGS`) and make 2 copies of code (one for `hashmap`, one for `hashmap_metrics`). `BUILD_METRICS` is still needed, I think.

Comment (tarik02) [reply CMakeLists.txt:5]: There should be ability to build metrics library in any case.

Comment (aminosbh) [reply CMakeLists.txt:5]: So, we should export also hashmap_metrics with an export name like HashMap::HashMapMetrics or HashMap::HashMap_metrics or HashMap::HashMap::Metrics

Comment (tarik02) [reply CMakeLists.txt:5]: I think, the third one is the best.

Comment (aminosbh) [reply CMakeLists.txt:5]: @Tarik02 It was a pleasure to discuss with you, and I hope you find it as instructive as it was for me. Actually, I will not be able to work on it these days, but I will get back to you as soon as possible.

Comment (tarik02) [reply CMakeLists.txt:5]: @aminosbh, thank you! It was interesting to me too. You may work on it as soon as you want.
